generator/processors/helpers.ts

```diff
@@ -2,9 +2,79 @@
 // Licensed under the MIT License.
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-export const replaceCyclicRef = (cyclicRef: any) => {
-  if (cyclicRef && cyclicRef['$ref']) {
-    delete cyclicRef['$ref'];
-    cyclicRef['type'] = 'object';
+export function replaceRefByName(schema: any, definitionName: string, replacement: any) {
+  doRecursively(schema, val => {
+    if (val['$ref'] === `#/definitions/${definitionName}`) {
+      return replacement;
+    }
+
+    return val;
+  })
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function replaceCyclicRefByName(schema: any, definitionName: string, replacement: any) {
+  const refVertices: Record<string, Set<string>> = {};
+
+  for (const refName in schema.definitions) {
+    refVertices[refName] = new Set<string>();
+
+    doRecursively(schema.definitions[refName], val => {
+      if (val['$ref'] && val['$ref'].startsWith('#/definitions/')) {
+        refVertices[refName].add(val['$ref'].substring('#/definitions/'.length));
+      }
+
+      return val;
+    });
+  }
+
+  const parentDefinitions = findCyclicReferencingDefinitions(refVertices, definitionName);
+  for (const parentDefinition of parentDefinitions) {
+    replaceRefByName(schema.definitions[parentDefinition], definitionName, replacement);
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function doRecursively(schema: any, func: (input: any) => any) {
```

Comment:
>doRecursively [](http://example.com/codeflow?start=9&length=13)

nit - `findDefinitionsRecursively` as the function name?

Reply:
Updated to `transformSchemaRecursively`.